로또 2단계 미션 제출합니다

Jummi10 opened the pull request:
자바 로또 1단계 제출합니다! 최대한 메서드와 클래스를 기능별로 나누고, 일급 컬렉션 적용해보려고 했는데 제대로 되었는지 확신이 서질 않네요. 부족한 점 많이 지적 부탁드려요!

그리고 질문이 있습니다
[`BonusNumber` 클래스](src/main/java/domain/value/BonusNumber.java)가 입력받은 `bonusNumber`값을 검증하는 `BonusNumberValidator`를 들고 있는데
`BonusNumberValidator`의 validate 메서드 중 하나가[ `WinningNumbers winningNumbers`](src/main/java/domain/value/WinningNumbers.java)가 필요해서 넘겨받을 수 있는 루트가 BonusNumber의 생성자 뿐이더라구요! 그래서
```java
 BonusNumber bonusNumber = new BonusNumber(receiver.receiveBonusNumber(), winningNumbers);
```
이런 식으로 winningNumbers도 생성자에 같이 넘겨주게 되었는데 혹시 이게 좋지 않은 방식인지 상관없는지 의견 들어보고 싶습니다

감사합니다 :)

Jummi10 committed:
docs: 기능 목록 작성

Jummi10 committed:
feat: 입력 구현

Jummi10 committed:
feat: 로또 도메인, 로또 번호 생성 구현

Jummi10 committed:
feat: 출력 메세지 분리

Jummi10 committed:
feat: 메세지 분리, 구입금액, 개수 입출력 구현

Jummi10 committed:
feat: 자동 번호 생성, 구입한 로또의 번호 출력

Jummi10 committed:
feat: 지난주 당첨 번호와 보너스 볼 입력 구현

Jummi10 committed:
feat: 총 당첨 금액과 총 수익률 계산 구현

- 일치하는 숫자의 개수에 따른 당첨금을 가지고 있는 enum

Jummi10 committed:
feat: 당첨 통계, 수익률 출력 구현

Jummi10 committed:
test: 당첨 확인 기능 구현 테스트 작성

Jummi10 committed:
chore: 패키지 구조 변경

- PrizeMoney -> WinningStatistics 으로 이름 변경

Jummi10 committed:
feat: 입력, 출력, 예외 메세지 분리, 구입금액 예외처리 구현

Jummi10 committed:
feat: 1-45 숫자인 targetNumbers 일급 컬렉션 적용

Jummi10 committed:
refactor: target number 45개 중 앞 6개 숫자 가져오는 역할 TargetNumbers에게 위임

Jummi10 committed:
feat: 로또 티켓 하나의 숫자를 당첨 번호, 보너스 번호와 비교하는 기능 구현

- 일치 개수와 그에 따른 상금을 가지고 있는 클래스 WinningStatistics -> PrizeMoney 로 변경
-

Jummi10 committed:
feat: 지난주 당첨 번호 보너스볼 입력 검증 구현

Jummi10 committed:
feat: 구매한 로또 티켓들과 당첨 번호를 비교하는 기능 구현

Jummi10 committed:
chore: 각 parameter명 수정, 숫자 상수화, 패키지 이동

Jummi10 committed:
feat: 각 티켓들의 결과로부터 당첨 통계 내는 기능 구현

- WinningStatistics 가 당첨 통계만 가지고 있도록 수정

Jummi10 committed:
chore: 패키지 구조 변경, Message 분리

Jummi10 committed:
refactor: 코드 병합

chae-heechan commented on README.md:
```diff
@@ -51,6 +51,42 @@
+### 예외
+
+- 구입 금액 입력
+  - [x] 구입금액이 0원 이하인 경우
```
구매 금액이 1000원 단위가 아닐때 예외 처리가 있었으면 좋겠어요!

chae-heechan commented on src/main/java/domain/lotteryStore/TargetNumbers.java:
```diff
@@ -0,0 +1,30 @@
+        for (int i = 1; i <= 45; i++) {
+            targetNumbers.add(i);
+        }
+    }
```
WinningNumbersValidator에서는 상수로 바꾸신것 같은데 여기서는 빠트리신것 같아요 상수로 빼주세요!

chae-heechan commented on src/main/java/domain/winningStatistics/PrizeMoney.java:
```diff
@@ -0,0 +1,37 @@
+    FOUR(4, 50000),
+    FIVE(5, 1500000),
+    FIVE_BONUS(5, 30000000),
+    SIX(6, 2000000000);
```
enum 사용 👍

chae-heechan commented on src/main/java/domain/value/WinningNumbers.java:
```diff
@@ -0,0 +1,47 @@
+public class WinningNumbers {
+    private static final String COMMA = ",";
+    private static final String BLANK = " ";
+    private static final String EMPTY_STRING = "";
```
다른 상수들 처럼 이 상수가 무슨역할을 하는지 이름에 써있으면 좋을 것 같아요

chae-heechan commented on src/test/java/domain/winningStatistics/WinningStatisticsTest.java:
```diff
@@ -0,0 +1,43 @@
+        assertThat(rankings.get(3)).isEqualTo(1);
+        assertThat(rankings.get(4)).isEqualTo(0);
+    }
+}
```
테스트 Printer, Receiver의 커버리지가 하나도 없는데 ByteArrayInputStream을 사용해서 Application을 한번 테스트하면 커버리지가 채워질 것 같아요!

kouz95 committed:
Update README.md

Jummi10 committed:
Merge branch 'main' of https://github.com/biforest/java-lotto into Jummi10

Jummi10 committed:
chore: 최소 최대 로또 번호 상수화

Jummi10 committed:
docs: 2단계 기능 목록 작성

Jummi10 committed:
feat: 로또 숫자 자동, 수동 생성 클래스 생성

- TargetNumbers -> 숫자 자동 생성으로 변경
- WinningNumbers -> 숫자 수동 생성으로 변경
- 상수명 변경
- validators 삭제

Jummi10 committed:
feat: 수동으로 구매할 로또 수와 번호 입력 구현

Jummi10 committed:
feat: 수동으로 입력받은 번호로 수동 로또 생성 구현

Jummi10 committed:
chore: 패키지 구조 변경, 당첨통계 메세지 변경, Application 리팩토링

Jummi10 committed:
refactor: 구매한 로또 번호와 지난주 당첨 번호 비교 로직을 Lotteries로 이동

Jummi10 committed:
refactor: PrizeMoney를 이용하여 당첨 통계 로직 구현

- 테스트 작성 및 수정

Jummi10 committed:
chore: end of file에 new line 추가, 메세지 수정

pkalsh commented:
> 자바 로또 1단계 제출합니다! 최대한 메서드와 클래스를 기능별로 나누고, 일급 컬렉션 적용해보려고 했는데 제대로 되었는지 확신이 서질 않네요. 부족한 점 많이 지적 부탁드려요!
> 
> 그리고 질문이 있습니다
> [`BonusNumber` 클래스](src/main/java/domain/value/BonusNumber.java)가 입력받은 `bonusNumber`값을 검증하는 `BonusNumberValidator`를 들고 있는데
> `BonusNumberValidator`의 validate 메서드 중 하나가[ `WinningNumbers winningNumbers`](src/main/java/domain/value/WinningNumbers.java)가 필요해서 넘겨받을 수 있는 루트가 BonusNumber의 생성자 뿐이더라구요! 그래서
> 
> ```java
>  BonusNumber bonusNumber = new BonusNumber(receiver.receiveBonusNumber(), winningNumbers);
> ```
> 
> 이런 식으로 winningNumbers도 생성자에 같이 넘겨주게 되었는데 혹시 이게 좋지 않은 방식인지 상관없는지 의견 들어보고 싶습니다
> 
> 감사합니다 :)

하고 계신 질문 자체에는 문제 없다고 생각합니다만 

보너스 번호를 VO로 포장하지 않고 당첨 번호 6개와 보너스 번호를 함께 포장하면 검증하는 메서드가 외부에서 어떠한 상태도 제공받지 않고 자율적으로 검증할 수 있지 않을까요?

또한, 개인적인 의견이지만 당첨 번호와 보너스 번호는 같은 성격을 공유하고 있기 때문에 같은 객체로 표현되어도 좋을 것 같습니다.

begaonnuri commented on README.md:
````diff
@@ -38,6 +38,60 @@
 총 수익률은 0.35입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)
 ```
 
+# 🚀 로또 2단계 - 수동 구매
````
기능목록을 상세하게 잘 작성해주셨네요 👍🏻

begaonnuri commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,47 @@
+    public static void main(String[] args) {
+        Application application = new Application();
+        application.run();
+    }
```
Application이 Application을 생성해서 실행시키는 것이 어색하게 느껴지네요.
run()은 애플리케이션의 흐름을 제어하면서 로직을 진행시키는 것 같은데 이것을 따로 객체로 분리해보면 어떨까요?
MVC 패턴과 Controller의 역할에 대해 학습해보면 좋을것같아요.

https://m.blog.naver.com/jhc9639/220967034588

begaonnuri commented on src/main/java/domain/lotteryStore/LotteryStore.java:
```diff
@@ -0,0 +1,46 @@
+import domain.lotteryStore.numbers.ManualNumbersGenerator;
+import domain.lotteryStore.numbers.Numbers;
+
+public class LotteryStore {
```
main에서 LotteryStore를 생성해서 사용하는데,
LotteryStore는 상태값을 갖고 있지 않기 때문에 굳이 객체가 생성될 필요가 없어보여요.
이런 경우엔 `createLotteries()`를 static하게 관리하는게 더 나아보여요.
어떻게 생각하시나요?

Jummi10 replied:
오호 네 static으로 바꾸는게 더 나아보입니다!

begaonnuri commented on src/main/java/domain/lotteryStore/LotteryStore.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private List<Lottery> makeAutoLotteries(int purchasedCount, int manualCount) {
+        AutoNumbersGenerator autoNumbersGenerator = new AutoNumbersGenerator();
```
여기서 객체를 생성하고 인자로 전달하는 이유가 있을까요?
`makeAutoLottery()`에서 생성되어도 괜찮을것같아서요.

Jummi10 replied:
`AutoNumbersGenerator`가 1-45의 숫자 더미를 가지고 생성되는데, `makeAutoLottery()`를 하며 하나의 자동 로또 티켓을 만들 때마다 매번 1-45를 만들 필요 없이
`makeAutoLotteries()`에서 한 번만 처음으로 `1-45`를 만들어두고 `makeAutoLottery()`에서는 shuffle()과 getSixNumbersFromTheFront()만 사용하고 싶어서 이렇게 autoNumbersGenerator를 인자로 전달하게 되었습니다!

이것보다 로또 하나를 생성할 때마다 AutoNumbersGenerator를 만드는게 더 나을까요??

begaonnuri commented on src/main/java/domain/lotteryStore/LotteryStore.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private Lottery makeAutoLottery(AutoNumbersGenerator autoNumbersGenerator) {
+        autoNumbersGenerator.shuffle();
+        Numbers numbers = new Numbers(autoNumbersGenerator.getSixNumbersFromTheFront());
+        return new Lottery(numbers);
```
autoNumbersGenerator를 shuffle하는 것은 AutoNumbersGenerator의 역할이 아닐까요?
그리고 `getSixNumbersFromTheFront()`은 항상 `shuffle()`이 선행되고 같이 사용되어야 할 것 같은데, 사용하는 측에선 이 부분을 모를것같아요.
shuffle하고 6개의 숫자를 리턴하는 로직을 AutoNumbersGenerator가 하도록 하면 어떨까요?

Jummi10 replied:
로또를 하나 만들 때 순서를 섞은 후 앞 6개의 숫자를 가져온다는 것을 순서대로 보여주고 싶어서 두 메서드를 모두 호출하였었는데, 말씀대로 두 메서드가 항상 동시에 시행되니 이를 AutoNumbersGenerator의 역할로 넘겨주어도 될 것 같네요!

begaonnuri commented on src/main/java/domain/lotteryStore/numbers/AutoNumbersGenerator.java:
```diff
@@ -0,0 +1,30 @@
+public class AutoNumbersGenerator {
+    private final List<Integer> autoNumbers = new ArrayList<>();
+
+    public AutoNumbersGenerator() {
+        initiateNumbers();
+    }
+
+    private void initiateNumbers() {
+        for (int i = LotteryNumberRange.MINIMUM_LOTTERY_NUMBER; i <= LotteryNumberRange.MAXIMUM_LOTTERY_NUMBER; i++) {
+            autoNumbers.add(i);
+        }
+    }
```
객체가 생성될 때 마다 매번 로또 넘버들을 생성하겠네요.
Java의 Integer 클래스는 일정 숫자들을 생성해두고 `캐싱`하는데요,
다음 링크를 참고해서 매번 같은 인스턴스를 반환하도록 변경해보는건 어떨까요?
https://antilog.tistory.com/63

Jummi10 replied:
오호 이 방식을 이용하면 [이 문제](https://github.com/biforest/java-lotto/pull/2#discussion_r613985121)가 해결되겠네요!

begaonnuri commented on src/main/java/domain/lotteryStore/numbers/LotteryNumberRange.java:
```diff
@@ -0,0 +1,18 @@
+
+import ui.message.ExceptionMessage;
+
+public class LotteryNumberRange {
+    public static final int MINIMUM_LOTTERY_NUMBER = 1;
+    public static final int MAXIMUM_LOTTERY_NUMBER = 45;
```
로또 숫자를 List<Integer> 타입으로 관리하는데, 생성된 로또를 검증해야 하다 보니 이 객체를 만드신 것 같아요.
그리고 이 코드가 중복되다 보니 NumberGenerator들이 상속받도록 한 것 같고요.

오늘 들으신 것 처럼 코드 재사용 만을 위한 상속은 지양하는 것이 좋습니다.
ManualNumberGenerator is a LotteryNumberRange가 어색하게 느껴진다면 적절한 상속의 사용이 아닐 수 있어요.

`로또 숫자는 1부터 45 사이에 있어야 한다.`
이것은 로또 숫자의 제약조건이고 책임이라고 볼 수 있을 것 같아요.
로또 숫자를 값 객체로 만들고 List<Integer>를 List<LottoNumber>와 같이 변경해보면 어떨까요?

Jummi10 replied:
> 로또 숫자는 1부터 45 사이에 있어야 한다.
이것은 로또 숫자의 제약조건이고 책임이라고 볼 수 있을 것 같아요.
로또 숫자를 값 객체로 만들고 List를 List와 같이 변경해보면 어떨까요?

이 부분을 잘 이해하지 못하였는데 int number 하나를 가진 로또 숫자를 객체로 만들고 숫자 하나의 validate()를 그 객체가 하게 하라는 말씀이시죠?? List를 List와 같이 변경하라는 말은 무슨 뜻인가요??

begaonnuri replied:
제가 말을 이상하게했었네요 😅

> 이 부분을 잘 이해하지 못하였는데 int number 하나를 가진 로또 숫자를 객체로 만들고 숫자 하나의 validate()를 그 객체가 하게 하라는 말씀이시죠??

네 맞습니다! 찰떡같이 알아들으시네요

begaonnuri commented on src/main/java/domain/lotteryStore/numbers/ManualNumbersGenerator.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private List<Integer> splitWinningNumbers(String input) {
+        return Arrays.stream(input.replaceAll(CHARACTER_BEFORE_REPLACING, CHARACTER_AFTER_REPLACING)
```
사용자를 배려해서 공백까지 제거해주셨네요 👍🏻 
```suggestion
        return Arrays.stream(input.trim()
```
trim()이란 api도 있답니다!
다양한 언어에서 이와 유사한 기능을 trim()이란 단어로 제공하니 알아두시면 좋을 것 같네요

Jummi10 replied:
trim()은 맨 앞 뒤에 있는 공백만 제거해주고 중간에 있는 공백은 제거해주지 못한다고 해서 replace를 사용하였습니다

begaonnuri commented on src/main/java/domain/lotteryStore/numbers/Numbers.java:
```diff
@@ -0,0 +1,28 @@
+public class Numbers {
+    private final List<Integer> numbers;
+
+    public Numbers(List<Integer> numbers) {
+        Collections.sort(numbers);
+        this.numbers = numbers;
+    }
```
Numbers가 sort되어있다는 것을 사용하는 측에선 알 수 없을 것 같아요.
SortedNumbers로 변경하거나 정적 팩토리 메소드를 사용해보는건 어떨까요?

그리고 애초에 정렬된 상태를 유지하는 것은 비즈니스 로직이 맞을까요?
정렬한 상태로 행해지는 기능이 없으면 출력만 정렬된 상태로 출력하도록 할 수도 있지 않을까요?
이 부분은 어떻게 생각하시는지 궁금하네요!

Jummi10 replied:
음 출력하는 부분에 sort 메서드를 넣는 것보다 Numbers 자체 내에서 sort를 하는 것이 더 낫다고 생각이 들어서 Numbers에 넣게 되었는데, 정렬된 상태를 사용하는 기능은 없다는 말에 공감이 되네요. 로또 티켓 하나가 가지고 있는 숫자들은 정렬된 순서가 없으니 sort되지 않은 numbers를 가지고, 출력할 때만 정렬을 해주어도 될 것 같습니다!

begaonnuri commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,47 @@
+        Lotteries lotteries = lotteryStore.createLotteries(purchasedCount, manualCount, manualNumbers);
+        printer.printPurchasedLotteries(lotteries);
+
+        ManualNumbersGenerator winningNumbers = new ManualNumbersGenerator(receiver.receiveWinningNumbers());
```
ManualNumbersGenerator가 winningNumbers인 것이 어색하게 느껴지네요.
winningNumbers의 타입은 Lottery 혹은 Numbers가 되는 것이 자연스럽지 않을까요?

begaonnuri commented on src/main/java/domain/lotteryStore/Lottery.java:
```diff
@@ -0,0 +1,15 @@
+
+import domain.lotteryStore.numbers.Numbers;
+
+public class Lottery {
+    private final Numbers numbers;
+
+    public Lottery(Numbers numbers) {
+        this.numbers = numbers;
+    }
+
+    public Numbers getNumbers() {
+        return numbers;
+    }
+}
```
Lottery가 단순히 값을 포장하는 용도로만 사용되고 있는데,
그렇다면 Numbers가 곧 Lottery가 아닐까요?
두 개를 구분하신 이유가 있을까요?

Jummi10 replied:
전체적으로 일급컬렉션을 적용하면서 구현하다가 List<Integer>를 포장하려고 하다보니 이렇게 나누게 되었는데, Lottery 자체가 numbers를 가지고 있다고 생각하면 굳이 더 나눌 필요가 없었네요..!
안 그래도 Lottery가 아무 역할이 없어서 고민이 되었었는데 그냥 Numbers의 역할을 Lottery로 다 넘겨주어야겠군요

begaonnuri commented on src/main/java/domain/winningStatistics/WinningStatistics.java:
```diff
@@ -0,0 +1,30 @@
+    private final Map<PrizeMoney, Integer> winningStatistics = new HashMap<>();
+
+    public WinningStatistics() {
+        initiate();
```
생성자가 복잡하지 않은데 메소드로 분리한 이유가 있을까요?

Jummi10 replied:
엇 엄청난 이유는 없고 그냥 작은 기능 하나하나를 모두 메서드로 나누고 만들려다보니 이렇게 되었습니다..😐

begaonnuri commented on src/main/java/domain/winningStatistics/PrizeMoney.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.Arrays;
+
+public enum PrizeMoney {
```
PrizeMoney란 이름이 적절하지 않은 것 같아요.
등 수, 당첨과 관련된 이름이 적절하지 않을까요?

begaonnuri commented on src/main/java/domain/winningStatistics/PrizeMoney.java:
```diff
@@ -0,0 +1,41 @@
+    SIX(6, false, 2000000000);
+
+    private final int matchingCount;
+    private final boolean isHavingBonusNumber;
```
지금도 나쁘진 않지만  `is-` 이외에도 `has-`도 boolean타입 필드명으로 사용되곤 합니다!

begaonnuri commented on src/main/java/domain/winningStatistics/PrizeMoney.java:
```diff
@@ -0,0 +1,41 @@
+
+    public static float calculateEarningsRate(WinningStatistics winningStatistics, int purchasedCount) {
+        int totalPrizeMoney = Arrays.stream(values())
+            .mapToInt(prizeMoney -> winningStatistics.get(prizeMoney) * prizeMoney.getPrizeMoney())
+            .sum();
+
+        return (float)totalPrizeMoney / (purchasedCount * 1000);
```
실수를 사용할땐 오차를 줄이기 위해 double 타입을 주로 사용합니다! 
https://hackr.io/blog/float-vs-double

begaonnuri replied:
이 1000이란 매직넘버는 로또 가격이 아닐까요? PurchasePrice의 상수를 import 할 수 있을 것 같네요.

begaonnuri commented on src/main/java/domain/winningStatistics/ComparisonResult.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private boolean isEqualMatchingCount(PrizeMoney prizeMoney) {
+        return matchingCount == prizeMoney.getMatchingCount();
+    }
+
+    private boolean isEqualBonusNumber(PrizeMoney prizeMoney) {
+        return isHavingBonusNumberWithFiveMatchingCount() == prizeMoney.isHavingBonusNumber();
+    }
+
+    private boolean isHavingBonusNumberWithFiveMatchingCount() {
+        return matchingCount == 5 && havingBonusNumber;
+    }
```
값이 꺼내져서(get) 비교되고 있는데, 
메소드 모두 `prizeMoney.isSameMatchCount(matchingCount)` 이런식으로 Enum에게 메시지를 던지는 형식으로 변경할 수 있을것 같네요.

begaonnuri commented on src/main/java/domain/winningStatistics/PrizeMoney.java:
```diff
@@ -0,0 +1,41 @@
+
+    public static float calculateEarningsRate(WinningStatistics winningStatistics, int purchasedCount) {
+        int totalPrizeMoney = Arrays.stream(values())
+            .mapToInt(prizeMoney -> winningStatistics.get(prizeMoney) * prizeMoney.getPrizeMoney())
```
굳이 getPrizeMoney를 하지 않아도 될 것 같은데 Enum의 이름과 필드의 이름이 동일해서 발생한 문제 같네요.
enum의 이름이 변경된다면 다음과 같이 사용해도 될 것 같아요.
```suggestion
            .mapToInt(바꾼이름 -> winningStatistics.get(바꾼이름) * prizeMoney)
```

begaonnuri reviewed (changes requested):
구현 잘 해주셨네요 👍🏻 
몇가지 변경사항 남겼으니 확인해주시고 코멘트가 이해 안되시면 DM주세요 :)

Jummi10 committed:
chore: 접근제어자 변경, 필드에 final 추가

Jummi10 committed:
refactor: Application에서 컨트롤러 분리 등 리뷰 반영

- AutoNumbersGenerator 1-45 캐싱
- Numbers 비즈니스 로직 -> Lottery로 이동 후 Numbers 삭제
- PrizeMoney -> Ranking 등수라는 것을 알 수 있게 수정
- 로또의 6개 숫자 정렬을 printer로 책임 이동

Jummi10 committed:
refactor: 정적 팩토리 메서드 사용

Jummi10 committed:
chore: 필드 삭제, BeforeEach로 초기화, 의미없는 test value source 삭제

Jummi10 committed:
refactor: 로또 티켓 하나의 번호를 지난주 당첨 번호, 보너스 번호와 비교하는 로직을 Lotteries -> Lottery에게 책임 이동

Jummi10 committed:
refactor: 자동 로또 생성에 shuffle strategy 적용

Jummi10 committed:
refactor: LotteryNumberRange -> LotteryNumber VO로 변경

- BonusNumber와 ManualNumbersGenerator의 상속 취소

begaonnuri commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,5 @@
+public class Application {
+    public static void main(String[] args) {
+        LottoController.run();
```
Controller가 static인 경우 여러 명의 사용자가 동시에 접근했을 때 문제가 생기지 않을까요?
이 경우엔 매번 controller를 생성하도록 하는건 어떨까요?

begaonnuri commented on src/main/java/LottoController.java:
```diff
@@ -0,0 +1,36 @@
+public class LottoController {
+    public static void run() {
+        final Receiver receiver = new Receiver();
+        final Printer printer = new Printer();
```
위의 리뷰와 더불어서 사실 controller는 매번 생성되어야 하지만 view에 관련된 것은 static하게 한번만 생성되어도 좋을것같네요!

begaonnuri commented on src/main/java/domain/lotteryStore/numbers/AutoShuffleStrategy.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+public class AutoShuffleStrategy implements ShuffleStrategy {
+    private AutoShuffleStrategy() {
+    }
+
+    public static AutoShuffleStrategy getInstance() {
+        return new AutoShuffleStrategy();
+    }
```
싱글톤 관리 👍🏻

begaonnuri reviewed (approved):
리뷰 반영해주시느라 수고 많으셨습니다! 💯 

코드가 충분히 깔끔해진 것 같아서 Approve 하겠습니다!
참고로 생각해보면 좋을 것들 몇가지 코멘트를 남겼으니 확인해주세요 :)

Jummi10 committed:
refactor: controller 메소드 static 삭제

- ui 부분 static하게 변경
